### PR TITLE
Jakarta JSON-P support.

### PR DESCRIPTION
**Pull request by paulogaspar7**
Add new JmesPath module to support Jakarta JSON-P 1.1.6.

It passes all tests.

### CODE REVIEW

**Comment by iconara** [comment]
Thank you for the contribution. I will look it through as soon as I can. Could you tell me more about your use case in the meantime? It's interesting to hear what people are using this library for.

**Comment by paulogaspar7** [comment]
A solution I am working on stores and serves JSON documents. I am playing with JMESPath for JSON transformation functionalities, both for document import / export as for creating alternate / partial document views. I think it can be composed with something like JSON Patch.

File: jmespath-jakarta-jsonp/src/main/java/io/burt/jmespath/jakarta/jsonp/JsonpRuntime.java [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+  private JsonValue nodeOrNullNode(JsonValue node) {
+    if (node == null) {
+      return JsonValue.NULL;
+    } else {
+      return node;
+    }
+  }
+
+  private String textOnStringOrNull(JsonValue value) {
```
Comment (iconara) [inline jmespath-jakarta-jsonp/src/main/java/io/burt/jmespath/jakarta/jsonp/JsonpRuntime.java]: This name confuses me, specifically the "on" part. Could it be called `stringOrNull` or `textOrNull`?

File: jmespath-jakarta-jsonp/src/main/java/io/burt/jmespath/jakarta/jsonp/JsonpRuntime.java [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+  @Override
+  public JsonValue createBoolean(boolean b) {
+    return b ? JsonValue.TRUE : JsonValue.FALSE;
+  }
+
+  @Override
+  public JsonValue createObject(Map<JsonValue, JsonValue> obj) {
+    JsonObjectBuilder builder = Json.createObjectBuilder();
+    for (Map.Entry<JsonValue, JsonValue> entry : obj.entrySet()) {
+      String key = textOnStringOrNull(entry.getKey());
```
Comment (iconara) [inline jmespath-jakarta-jsonp/src/main/java/io/burt/jmespath/jakarta/jsonp/JsonpRuntime.java]: Can `key` really be null here? That's as far as I know not allowed in JSON, and the docs for the `createObject` method in `Adapter` says "The map keys must all be string values."

The Jackson adapter does not check for null in its implementation of this method so it's safe to skip the check.

File: jmespath-jakarta-jsonp/src/main/java/io/burt/jmespath/jakarta/jsonp/JsonpRuntime.java [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        return JmesPathType.NUMBER;
+      default:
+        throw new IllegalStateException(String.format("Unknown node type encountered: %s", value.getValueType()));
+    }
+  }
+
+  @Override
+  public JsonValue getProperty(JsonValue value, JsonValue name) {
+    if (value.getValueType() == OBJECT) {
+      return nodeOrNullNode(((JsonObject)value).get(textOnStringOrNull(name)));
```
Comment (iconara) [inline jmespath-jakarta-jsonp/src/main/java/io/burt/jmespath/jakarta/jsonp/JsonpRuntime.java]: How come the call to `textOnStringOrNull` here? The object can't have a null key. I think `name.getString()` should be sufficient here.

File: jmespath-jakarta-jsonp/src/main/java/io/burt/jmespath/jakarta/jsonp/JsonpRuntime.java (hunk 1 above)
Comment (iconara) [reply jmespath-jakarta-jsonp/src/main/java/io/burt/jmespath/jakarta/jsonp/JsonpRuntime.java]: I've looked at the places where this method is used, and I think it's in fact not needed. The callers (`createObject` and `getProperty`) don't need to call this method.

File: jmespath-jakarta-jsonp/src/main/java/io/burt/jmespath/jakarta/jsonp/JsonpRuntime.java [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+  public JsonValue createNumber(double n) {
+    return Json.createValue(n);
+  }
+
+  @Override
+  public JsonValue createNumber(long n) {
+    return Json.createValue(n);
+  }
+
+  private JsonValue useJsonNull(JsonValue value) { return (value == null) ? JsonValue.NULL : value; }
```
[the remark this answers is not in the document]
Comment (paulogaspar7) [reply jmespath-jakarta-jsonp/src/main/java/io/burt/jmespath/jakarta/jsonp/JsonpRuntime.java]: I can't believe I did this!

File: jmespath-jakarta-jsonp/src/main/java/io/burt/jmespath/jakarta/jsonp/JsonpRuntime.java (hunk 1 above)
Comment (paulogaspar7) [reply jmespath-jakarta-jsonp/src/main/java/io/burt/jmespath/jakarta/jsonp/JsonpRuntime.java]: I replaced it by the toString(key) call.

Comment (paulogaspar7) [reply jmespath-jakarta-jsonp/src/main/java/io/burt/jmespath/jakarta/jsonp/JsonpRuntime.java]: (Still being a bit paranoid here.)

**Comment by paulogaspar7** [comment]
Suggested changes applied. Please check if it is ok now.

File: jmespath-jakarta-jsonp/src/main/java/io/burt/jmespath/jakarta/jsonp/JsonpRuntime.java (hunk 3 above)
Comment (paulogaspar7) [reply jmespath-jakarta-jsonp/src/main/java/io/burt/jmespath/jakarta/jsonp/JsonpRuntime.java]: >  it's very nice to see how it's possible to support another library with this little effort.

Indeed! It just took some hours (part of a day) and most of the effort was spent figuring out the right semantics of this API.

... and I do not benefit from that much experience with JSON, which kind of shows on the glitches you pointed out.

The test battery helps a lot!!!

Comment (paulogaspar7) [reply jmespath-jakarta-jsonp/src/main/java/io/burt/jmespath/jakarta/jsonp/JsonpRuntime.java]: Congratulations. This library is a really easy both to use and to code for!
:)

**Comment by iconara** [comment]
I went ahead and merged. I'll post here when I've done the release, it be a day or two, sorry.

**Comment by paulogaspar7** [comment]
Thank you Theo, that is just fine.

**Comment by iconara** [comment]
Released as [v0.5.0](https://github.com/burtcorp/jmespath-java/releases/tag/v0.5.0). Sorry for the delay, and thanks again for the contribution!

**Comment by paulogaspar7** [comment]
No problem. It was short enough for me. :)
Thank you for your really nice and very useful work.

**Pull request merged**